core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java

```diff
@@ -0,0 +1,189 @@
+        this.codec = codec;
+    }
+
+    /**
+     * @param codec codec name
+     * @param blobNameFormat format of the blobname in {@link String#format} format
+     * @param reader         prototype object that can deserialize T from XContent
+     * @param wrapper        true if content should be surrounded with extra start/endObject during serialization
+     * @param compress      true if the content should be compressed
+     */
+    public ChecksumBlobStoreFormat(String codec, String blobNameFormat, FromXContentBuilder<T> reader, ParseFieldMatcher parseFieldMatcher, boolean wrapper, boolean compress) {
+        this(codec, blobNameFormat, reader, parseFieldMatcher, wrapper, compress, DEFAULT_X_CONTENT_TYPE);
+    }
+
+    public T readBlob(BlobContainer blobContainer, String blobName) throws IOException {
+        try (InputStream inputStream = blobContainer.openInput(blobName)) {
+            byte[] bytes = ByteStreams.toByteArray(inputStream);
+            final String resourceDesc = "ChecksumBlobStoreFormat.readBlob(blob=\"" + blobName + "\")";
+            try (ByteArrayIndexInput indexInput = new ByteArrayIndexInput(resourceDesc, bytes)) {
+                CodecUtil.checksumEntireFile(indexInput);
+                CodecUtil.checkHeader(indexInput, codec, VERSION, VERSION);
+                long filePointer = indexInput.getFilePointer();
+                long contentSize = indexInput.length() - CodecUtil.footerLength() - filePointer;
+                BytesReference bytesReference = new BytesArray(bytes, (int)filePointer, (int)contentSize);
+                return read(bytesReference);
+            } catch(CorruptIndexException | IndexFormatTooOldException | IndexFormatTooNewException ex) {
+                // we trick this into a dedicated exception with the original stacktrace
+                throw new CorruptStateException(ex);
+            }
+        }
+    }
+
+    /**
+     * Writes properly formatted blob in atomic manner.
+     * <p/>
+     * The blob will be compressed and checksum will be written if required.
+     *
+     * @param obj           object to be serialized
+     * @param blobContainer blob container
+     * @param name          blob name
+     * @throws IOException
+     */
+    public void writeAtomic(T obj, BlobContainer blobContainer, String name) throws IOException {
+        String blobName = blobName(name);
+        String tempBlobName = tempBlobName(name);
+        writeBlob(obj, blobContainer, tempBlobName);
+        blobContainer.move(tempBlobName, blobName);
+    }
+
+    /**
+     * Writes properly formatted blob in atomic manner.
+     * <p/>
+     * The blob will be compressed and checksum will be written if required.
+     *
+     * @param obj           object to be serialized
+     * @param blobContainer blob container
+     * @param name          blob name
+     * @throws IOException
+     */
+    public void write(T obj, BlobContainer blobContainer, String name) throws IOException {
```

Comment:
I tried to do that but it creates a lot of inefficiency on slow repositories and problems while trying to create a snapshot with the same name as previously failed snapshot. We can do this switch, but I would prefer move it into another PR.